pkg/oauthflow/pkce.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright © 2021 Bob Callaway <bcallawa@redhat.com>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package oauthflow
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"golang.org/x/oauth2"
+
+	"github.com/sigstore/sigstore/pkg/log"
+)
+
+const (
+	PKCES256 = "S256"
+)
+
+type PKCE struct {
+	Challenge string
+	Method    string
+	Value     string
+}
+
+func NewPKCE(provider *oidc.Provider) (*PKCE, error) {
+	var providerClaims struct {
+		CodeChallengeMethodsSupported []string `json:"code_challenge_methods_supported"`
+	}
+
+	if err := provider.Claims(&providerClaims); err != nil {
+		// will only error out if the JSON was malformed, which shouldn't happen at this point
+		return nil, err
+	}
+
+	var chosenMethod string
+	for _, method := range providerClaims.CodeChallengeMethodsSupported {
+		// per RFC7636, any server that supports PKCE must support S256
+		if method == PKCES256 && chosenMethod != PKCES256 {
+			chosenMethod = PKCES256
+			break
+		} else if method != "plain" {
+			log.Logger.Infof("Unsupported code challenge method in list: '%v'", method)
```

Comment:
I know you just copied this over, but we might want to use the cli.Logger here rather than the production one.

Reply:
yup, totally. makes sense